zeroconf/__init__.py

```diff
@@ -2894,14 +2943,24 @@ def remove_listener(self, listener: RecordUpdateListener) -> None:
         except Exception as e:  # pylint: disable=broad-except  # TODO stop catching all Exceptions
             log.exception('Unknown error, possibly benign: %r', e)
 
-    def update_record(self, now: float, rec: DNSRecord) -> None:
+    @contextmanager
+    def update_records(self, now: float, rec: List[DNSRecord]) -> Generator:
         """Used to notify listeners of new information that has updated
-        a record."""
-        for listener in self.listeners:
-            listener.update_record(self, now, rec)
-        self.notify_all()
+        a record.
+
+        This method must be called before the cache is updated.
+        """
+        try:
+            for listener in self.listeners:
+                listener.update_records(self, now, rec)
+            self.notify_all()
```

Comment:
I think we can drop this notify once https://github.com/jstasiak/python-zeroconf/issues/476 is fixed